### PR TITLE
chore(schema): configure traffic source rules

### PR DESCRIPTION
**Pull request by tyyzqmf**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

截屏2024-04-09 20 26 21

### s3 files info: 
@amliuyong 
```
const bucketPrefix = `clickstream/${projectId}/rules/${appId}/`;
export const TRAFFIC_SOURCE_CATEGORY_RULE_FILE_NAME = 'traffic_source_category_rule_v1.json';
export const TRAFFIC_SOURCE_CHANNEL_RULE_FILE_NAME = 'traffic_source_channel_rule_v1.json';
```

## Test checklist

- [x] add new test cases
- [ ] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [ ] deploy data modeling
    - [ ] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting
  - [ ] streaming ingestion
    - [ ] with Redshift Serverless
    - [ ] with provisioned Redshift

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

File: src/control-plane/backend/lambda/api/store/aws/s3.ts [hunk 1]
```diff
@@ -125,3 +100,61 @@ export const isBucketExist = async (region: string, bucket: string) => {
...
   }
 };
+
+export async function putStringToS3(
+  content: string,
+  bucketName: string,
+  key: string,
+) {
+  const s3Client = new S3Client({
+    ...aws_sdk_client_common_config,
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/store/aws/s3.ts:110]: specify the region for opt-in regions for s3 client

**Review by zxkane — CHANGES_REQUESTED** [review]

**Review by zxkane — CHANGES_REQUESTED** [review]
- [ ] resolve conflicts locally

File: frontend/public/locales/zh-CN/analytics.json [hunk 2]
```diff
@@ -192,6 +199,48 @@
         "title": "用户属性详细信息"
       }
     },
+    "trafficSource": {
+      "title": "流量来源",
+      "alert": "对流量源设置的任何更改都只适用于未来的数据，历史数据不会被重新分类。",
+      "channelGroup": {
+        "title": "渠道组",
+        "description": "渠道组是一组渠道，这些渠道是基于规则的流量来源类别。Clickstream解决方案提供了一个默认的渠道组来对流量源进行分类，但您可以灵活地进行自定义。",
```
Comment (luorobin-a2z) [inline frontend/public/locales/zh-CN/analytics.json:207]: Clickstream --> 点击流

**Review by luorobin-a2z — DISMISSED** [review]
small edits

File: frontend/public/locales/zh-CN/help.json [hunk 3]
```diff
@@ -139,13 +139,5 @@
       "docLinkName": "事件属性",
       "docLink": "/analytics/data-mgmt/#_6"
     }
-  },
-  "metadataUserParamInfo": {
```
[the remark this answers is not in the document]
Comment (tyyzqmf) [reply frontend/public/locales/zh-CN/help.json:143]: reset it.

File: frontend/src/pages/analytics/traffic-source/ChannelGroup.tsx [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+    },
+    {
+      id: 'actions',
+      header: t('analytics:metadata.trafficSource.channelGroup.columnActions'),
+      cell: (item: IChannelGroup) => cellRenderer(item),
+      width: 100,
+    },
+  ];
+
+  const CONTENT_DISPLAY = [
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/traffic-source/ChannelGroup.tsx:276]: Move this out of components

File: frontend/src/pages/analytics/traffic-source/TrafficSourceHome.tsx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  useEffect(() => {
+    fetchTrafficSource();
+  }, []);
+
+  return (
+    <>
+      <Alert statusIconAriaLabel="Info">
+        {t('analytics:metadata.trafficSource.alert')}
+      </Alert>
+      <br />
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/traffic-source/TrafficSourceHome.tsx:74]: Replace `<br />` with `<SpaceBetween direction="vertical" size="l">` Component

File: frontend/src/pages/analytics/traffic-source/modal/ChannelGroupModal.tsx [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+  }, [selectedItems, modalType]);
+
+  return (
+    <>
+      <Modal
+        onDismiss={() => {
+          setVisible(false);
+        }}
+        visible={visible}
+        closeAriaLabel="Close modal"
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/traffic-source/modal/ChannelGroupModal.tsx:174]: Remove this if no need or if no i18n

File: frontend/src/pages/analytics/traffic-source/modal/ChannelGroupModal.tsx [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+                {t('button.cancel')}
+              </Button>
+              <Button
+                variant="primary"
+                loading={loading}
+                onClick={() => {
+                  if (checkInput()) {
+                    if (
+                      modalType ===
+                      t('analytics:metadata.trafficSource.modalType.edit')
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/traffic-source/modal/ChannelGroupModal.tsx:194]: Why use i18n result for condition? Why not enum type?

File: frontend/src/pages/analytics/traffic-source/modal/ChannelGroupModal.tsx [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+                  }
+                }}
+              >
+                {t('button.confirm')}
+              </Button>
+            </SpaceBetween>
+          </Box>
+        }
+        header={
+          defaultStr(t('analytics:metadata.trafficSource.channelGroup.title')) +
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/traffic-source/modal/ChannelGroupModal.tsx:209]: use `abc ${xxx} bcd` (Template literals ([Template strings](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals)))

File: frontend/src/pages/analytics/traffic-source/modal/ChannelGroupModal.tsx [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+        <SpaceBetween direction="vertical" size="m">
+          <FormField
+            label={t(
+              'analytics:metadata.trafficSource.channelGroup.columnName'
+            )}
+            description={t(
+              'analytics:metadata.trafficSource.channelGroup.nameDesc'
+            )}
+            errorText={
+              inputNameError
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/traffic-source/modal/ChannelGroupModal.tsx:223]: use `ternary` function from `utils.ts` for reduce complexity.

File: frontend/src/pages/analytics/traffic-source/modal/ChannelGroupModal.tsx [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+              )}
+              value={newDescription}
+            />
+          </FormField>
+          <FormField
+            label={t(
+              'analytics:metadata.trafficSource.channelGroup.columnCondition'
+            )}
+            errorText={
+              inputConditionError
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/traffic-source/modal/ChannelGroupModal.tsx:263]: use `ternary` function from `utils.ts` for reduce complexity.

File: frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+
+  const [newSource, setNewSource] = useState('');
+  const [inputSourceError, setInputSourceError] = useState(false);
+
+  const [selectedCategory, setSelectedCategory] =
+    useState<SelectProps.Option | null>(null);
+  const [selectedCategoryError, setSelectedCategoryError] = useState(false);
+
+  const [newPattern, setNewPattern] = useState<string>('');
+  const [newPatterns, setNewPatterns] = useState<{ label: string }[]>([]);
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx:78]: `{label: string} `means `SelectProps.Option`?

File: frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx [hunk 12]
```diff
@@ -0,0 +1,364 @@
...
+  const [inputPatternNumError, setInputPatternNumError] = useState(false);
+
+  const categoryOptions = [
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx:81]: Move this out of component

File: frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx [hunk 13]
```diff
@@ -0,0 +1,364 @@
...
+        categoryOptions.find((item) => item.value === selectedItem.category) ||
+          null
+      );
+      setNewPatterns(
+        selectedItem.params.map((item) => {
+          return { label: item };
+        })
+      );
+    }
+    if (modalType === t('analytics:metadata.trafficSource.modalType.new')) {
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx:190]: Why use i18n result as condition?

File: frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+              >
+                {t('button.cancel')}
+              </Button>
+              <Button
+                variant="primary"
+                loading={loading}
+                onClick={() => {
+                  if (checkInput()) {
+                    if (
+                      modalType ===
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx:221]: Why use i18n result as condition?

File: frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+                  }
+                }}
+              >
+                {t('button.confirm')}
+              </Button>
+            </SpaceBetween>
+          </Box>
+        }
+        header={
+          defaultStr(
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx:237]: change to  Template literals ([Template strings](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals))

File: frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx [hunk 16]
```diff
@@ -0,0 +1,364 @@
...
+        <SpaceBetween direction="vertical" size="m">
+          <FormField
+            label={t(
+              'analytics:metadata.trafficSource.sourceCategory.columnDomain'
+            )}
+            description={t(
+              'analytics:metadata.trafficSource.sourceCategory.domainDesc'
+            )}
+            errorText={
+              inputDomainError ? t('analytics:valid.inputDomainError') : ''
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx:253]: use ternary

File: frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx [hunk 17]
```diff
@@ -0,0 +1,364 @@
...
+          </FormField>
+          <FormField
+            label={t(
+              'analytics:metadata.trafficSource.sourceCategory.columnName'
+            )}
+            description={t(
+              'analytics:metadata.trafficSource.sourceCategory.sourceDesc'
+            )}
+            errorText={
+              inputSourceError ? t('analytics:valid.inputSourceError') : ''
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx:275]: use ternary

File: frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx [hunk 18]
```diff
@@ -0,0 +1,364 @@
...
+          </FormField>
+          <FormField
+            label={t(
+              'analytics:metadata.trafficSource.sourceCategory.columnCategory'
+            )}
+            description={t(
+              'analytics:metadata.trafficSource.sourceCategory.categoryDesc'
+            )}
+            errorText={
+              selectedCategoryError
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx:297]: use ternary

File: frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx [hunk 19]
```diff
@@ -0,0 +1,364 @@
...
+          </FormField>
+          <FormField
+            label={t(
+              'analytics:metadata.trafficSource.sourceCategory.columnPattern'
+            )}
+            description={t(
+              'analytics:metadata.trafficSource.sourceCategory.patternDesc'
+            )}
+            errorText={
+              inputPatternNumError
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx:322]: use ternary

File: frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx (hunk 14 above)
Comment (tyyzqmf) [reply frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx:221]: Add new Enum type

File: frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx (hunk 11 above)
Comment (tyyzqmf) [reply frontend/src/pages/analytics/traffic-source/modal/SourceCategoryModal.tsx:78]: `TokenGroupProps.Item`

File: src/control-plane/backend/lambda/api/store/aws/s3.ts [hunk 20]
```diff
@@ -125,3 +100,63 @@ export const isBucketExist = async (region: string, bucket: string) => {
...
+      Key: key,
+      Body: content,
+    }),
+  );
+}
+
+export async function readS3ObjectAsString(bucketName: string, key: string): Promise<string | undefined> {
+  try {
+    const s3Client = new S3Client({
+      ...aws_sdk_client_common_config,
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/store/aws/s3.ts:126]: miss `region` for working on opt-in region

**Review by zxkane — CHANGES_REQUESTED** [review]

**Review by chenhaiyun — APPROVED** [review]

File: src/control-plane/backend/lambda/api/store/aws/s3.ts (hunk 20 above)
Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/store/aws/s3.ts:126]: done

**Review by zxkane — APPROVED** [review]

**Pull request merged**